File: epic_ops/src/epic_ops/expand.cpp
#include <torch/library.h>
#include "epic_ops/expand.h"

namespace epic_ops::expand {

std::tuple<at::Tensor, at::Tensor> expand_csr(const at::Tensor& offsets, int64_t output_size) {
  static auto op = c10::Dispatcher::singleton()
                       .findSchemaOrThrow("epic_ops::expand_csr", "")
                       .typed<decltype(expand_csr)>();
  return op.call(offsets, output_size);
}

TORCH_LIBRARY_FRAGMENT(epic_ops, m) {
  m.def(TORCH_SELECTIVE_SCHEMA(
      "epic_ops::expand_csr(Tensor offsets, int output_size) -> (Tensor, Tensor)"));
}

} // namespace epic_ops::expand

File: epic_ops/epic_ops/expand.py
from typing import Tuple

import torch


@torch.no_grad()
def expand_csr(
    offsets: torch.Tensor,
    output_size: int,
) -> Tuple[torch.Tensor, torch.Tensor]:
    offsets = offsets.contiguous()

    return torch.ops.epic_ops.expand_csr(
        offsets, output_size
    )


def test():
    offsets = torch.as_tensor([
        # 0, 7, 9, 20
        0, 4, 8
    ], dtype=torch.int64, device="cuda")

    result, sizes = expand_csr(offsets, 20)
    print("result: ", result)
    print("sizes: ", sizes)


if __name__ == "__main__":
    test()
